python/ray/data/_internal/planner/plan_udf_map_op.py

```diff
@@ -12,73 +22,186 @@
-        assert isinstance(compute, ActorPoolStrategy)
-
         fn_constructor_args = op._fn_constructor_args or ()
         fn_constructor_kwargs = op._fn_constructor_kwargs or {}
 
         fn_ = make_callable_class_concurrent(op._fn)
 
-        def fn(item: Any, *args, **kwargs) -> Any:
+        def fn(item: Any) -> Any:
             assert ray.data._cached_fn is not None
             assert ray.data._cached_cls == fn_
-            return ray.data._cached_fn(item, *args, **kwargs)
+            return ray.data._cached_fn(item, *fn_args, **fn_kwargs)
 
         def init_fn():
             if ray.data._cached_fn is None:
                 ray.data._cached_cls = fn_
                 ray.data._cached_fn = fn_(*fn_constructor_args, **fn_constructor_kwargs)
 
     else:
-        fn = op._fn
-        init_fn = None
-    fn_args = (fn,)
-    if op._fn_args:
-        fn_args += op._fn_args
-    fn_kwargs = op._fn_kwargs or {}
 
-    def do_map(blocks: Iterator[Block], ctx: TaskContext) -> Iterator[Block]:
-        yield from transform_fn(blocks, ctx, *fn_args, **fn_kwargs)
+        def fn(item: Any) -> Any:
+            return op._fn(item, *fn_args, **fn_kwargs)
+
+        def init_fn():
+            pass
+
+    return fn, init_fn
+
+
+def validate_batch(batch: Block) -> None:
+    if not isinstance(
+        batch,
+        (
+            list,
+            pa.Table,
+            np.ndarray,
+            collections.abc.Mapping,
+            pd.core.frame.DataFrame,
+        ),
+    ):
+        raise ValueError(
+            "The `fn` you passed to `map_batches` returned a value of type "
+            f"{type(batch)}. This isn't allowed -- `map_batches` expects "
+            "`fn` to return a `pandas.DataFrame`, `pyarrow.Table`, "
+            "`numpy.ndarray`, `list`, or `dict[str, numpy.ndarray]`."
+        )
+
+    if isinstance(batch, list):
+        raise ValueError(
+            f"Error validating {_truncated_repr(batch)}: "
+            "Returning a list of objects from `map_batches` is not "
+            "allowed in Ray 2.5. To return Python objects, "
+            "wrap them in a named dict field, e.g., "
+            "return `{'results': objects}` instead of just `objects`."
+        )
+
+    if isinstance(batch, collections.abc.Mapping):
+        for key, value in list(batch.items()):
+            if not is_valid_udf_return(value):
+                raise ValueError(
+                    f"Error validating {_truncated_repr(batch)}: "
+                    "The `fn` you passed to `map_batches` returned a "
+                    f"`dict`. `map_batches` expects all `dict` values "
+                    f"to be `list` or `np.ndarray` type, but the value "
+                    f"corresponding to key {key!r} is of type "
+                    f"{type(value)}. To fix this issue, convert "
+                    f"the {type(value)} to a `np.ndarray`."
+                )
+
+
+def _create_map_data_processor_for_map_batches_op(op: MapBatches):
```

Comment:
can we put this function and related functions into a separate file, such as the original `map_batches.py`?

Can we call it `_generate_composer_for_map_batches_op`?

Reply:
I actually intentionally wanted to move these functions to this file. Because I feel it kind of hurt readability to have so many small files. And also should code that could be shared gets duplicated because of this. 
Is there any particular reason why you prefer put them in separate files?

Reply:
By the way, for those generate_xxx functions, I think I can just change them to use the new MapDataProcessor.

Reply:
I was actually thinking to put the functions for each operator in separate files, so we can swap the implementation easily, and code can easily get complicated. For context, all these code was defined in `dataset.py` before the refactoring. If you want to put all code into a single file, can we put all of them into a separate file, such as `map_functions.py`/`map_ops_functions`?